python/ray/scripts/symmetric_run.py

```diff
@@ -0,0 +1,194 @@
+"""Symmetric Run for Ray.
+
+This script is used to run a command on all nodes in a Ray cluster. Most useful for HPC settings.
+
+Example:
+
+    python -m ray.scripts.symmetric_run --address 127.0.0.1:6379 -- python my_script.py
+
+    xpanes -c 4 "python -m ray.scripts.symmetric_run --address 127.0.0.1:6379 --nnodes 4 -- python my_script.py"
+
+
+"""
+
+from typing import List
+
+import click
+import ray
+import socket
+import os
+import subprocess
+import sys
+import time
+from urllib.parse import urlparse
+
+import psutil
+
+CLUSTER_WAIT_TIMEOUT = os.environ.get("RAY_SYMMETRIC_RUN_CLUSTER_WAIT_TIMEOUT", 30)
+
+
+def check_cluster_ready(nnodes, timeout=CLUSTER_WAIT_TIMEOUT):
+    """Wait for all nodes to start.
+
+    Raises an exception if the nodes don't start in time.
+    """
+    start_time = time.time()
+    current_nodes = 1
+    ray.init(ignore_reinit_error=True)
+
+    while time.time() - start_time < timeout:
+        time.sleep(5)
+        current_nodes = len(ray.nodes())
+        if current_nodes == nnodes:
+            return True
+        else:
+            click.echo(
+                f"Waiting for nodes to start... {current_nodes}/{nnodes} nodes started"
+            )
+    return False
+
+
+def is_port_open(host: str, port: int, timeout: int = 5) -> bool:
+    """Check if a port is open on a given host."""
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.settimeout(timeout)
+            result = s.connect_ex((host, port))
+            return result == 0
+    except Exception:
+        return False
+
+
+def check_head_node_ready(head_ip, head_port, timeout=CLUSTER_WAIT_TIMEOUT):
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        if is_port_open(head_ip, head_port):
+            print("Ray cluster is ready!")
+            return True
+        time.sleep(5)
+
+    print(f"Timeout: Ray cluster at {head_ip}:{head_port} not ready after {timeout}s")
+    return False
+
+
+@click.command()
+@click.option(
+    "--address", required=True, type=str, help="The address of the Ray cluster."
+)
+@click.option(
+    "--nnodes", required=True, type=int, help="The number of nodes in the cluster."
+)
```

Comment:
hm, is this strictly required? we should be able to have worker nodes join as they come up and the script can wait for them all if desired?

Reply:
yup, changed